// pry/lib/pry/sync_prompt.js
(function() {
  var EventEmitter, MultilineState, ReverseSearch, SinglelineState, SyncPrompt, chalk, deasync, fs, readline,
    bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  readline = require('readline');

  fs = require('fs');

  EventEmitter = (require('events')).EventEmitter;

  deasync = require('deasync');

  chalk = require('chalk');

  MultilineState = (function() {
    function MultilineState() {}

    MultilineState.prototype.data = '';

    MultilineState.prototype.keypress = function(input, chars) {
      this.data += chars;
      if (this.data.match(/(\r|\n)\1$/)) {
        this.data = '';
        input.state('single');
        return input.send_data();
      } else if (chars.match(/(\r|\n)$/)) {
        return input.prompt();
      }
    };

    MultilineState.prototype.prompt = function(input, prompt) {
      if (this.data === '') {
        input.cli.setPrompt(prompt.replace(/[^>](?!$)/g, '-'));
      } else {
        input.cli.setPrompt(prompt.replace(/.(?!$)/g, '.'));
      }
      return input.cli._refreshLine();
    };

    return MultilineState;

  })();

  SinglelineState = (function() {
    function SinglelineState() {}

    SinglelineState.prototype.keypress = function(input, chars) {
      if (chars === '') {
        input.state('search');
        return input.prompt();
      } else if (chars === '\u0016') {
        input.state('multi');
        return input.prompt();
      } else if (chars.match(/(\r|\n)$/)) {
        return input.send_data();
      }
    };

    SinglelineState.prototype.prompt = function(input, prompt) {
      input.cli.setPrompt(prompt);
      return input.cli.prompt();
    };

    return SinglelineState;

  })();

  ReverseSearch = (function() {
    function ReverseSearch() {}

    ReverseSearch.prototype.init = function() {
      return this.index = -1;
    };

    ReverseSearch.prototype.quit = function(input) {
      input.state('single');
      input.prompt();
      input.cli._moveCursor(input.cli.line.length);
      return input.cli._refreshLine();
    };

    ReverseSearch.prototype.keypress = function(input, chars) {
      var history, i;
      if (chars === '') {
        return this.quit(input);
      }
      if (chars.charCodeAt() === 13) {
        if (this.searchResult) {
          input.cli.history.shift();
          input.cli.history.unshift(this.searchResult);
          input.type(this.searchResult);
        }
        return this.quit(input);
      }
      i = chars === '' ? this.index : -1;
      history = input.cli.history;
      while (i < history.length - 1) {
        if (history[++i].indexOf(input.cli.line) > 0) {
          this.index = i;
          break;
        }
      }
      this.searchResult = history[this.index];
      return input.prompt();
    };

    ReverseSearch.prototype.prompt = function(input, prompt) {
      var pre;
      if (this.searchResult === null && input.cli.line !== '') {
        pre = 'failed-';
      }
      input.cli.setPrompt("(" + (pre || '') + "reverse-i-search)`" + input.cli.line + "': " + (this.searchResult || '') + "      # ");
      return input.cli._refreshLine();
    };

    return ReverseSearch;

  })();

  SyncPrompt = (function(superClass) {
    extend(SyncPrompt, superClass);

    SyncPrompt.prototype.lines = '';

    SyncPrompt.prototype.count = 0;

    SyncPrompt.prototype.done = false;

    SyncPrompt.prototype._state = 'single';

    SyncPrompt.prototype.states = {
      multi: new MultilineState,
      single: new SinglelineState,
      search: new ReverseSearch
    };

    function SyncPrompt(arg) {
      var error, hist, lastLine, typeahead;
      typeahead = arg.typeahead, this.mode = arg.mode;
      this.close = bind(this.close, this);
      this.type = bind(this.type, this);
      this.prompt = bind(this.prompt, this);
      this.send_data = bind(this.send_data, this);
      this.keypress = bind(this.keypress, this);
      this.line = bind(this.line, this);
      this.state = bind(this.state, this);
      SyncPrompt.__super__.constructor.apply(this, arguments);
      this.indent = '';
      this.cli = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
        completer: typeahead,
        terminal: true,
        historySize: 500
      });
      try {
        hist = fs.readFileSync(process.env.HOME + "/.pryjs_history").toString();
        this.cli.history = hist.split('\n').reverse();
      } catch (error) {
        null;
      }
      lastLine = this.cli.history[0];
      this.cli.on('line', (function(_this) {
        return function(line) {
          if (line && line.length && line !== lastLine) {
            fs.appendFile(process.env.HOME + "/.pryjs_history", "\n" + line, function() {});
            lastLine = line;
          }
          return _this.line(line);
        };
      })(this));
      process.stdin.on('data', this.keypress);
    }

    SyncPrompt.prototype.state = function(state) {
      var base;
      if (state) {
        this._state = state;
        if (typeof (base = this.states[state]).init === "function") {
          base.init();
        }
      }
      return this.states[this._state];
    };

    SyncPrompt.prototype.line = function(line) {
      if (line.charCodeAt(0) === 22) {
        line = line.slice(1);
      }
      return this.lines += '\n' + line;
    };

    SyncPrompt.prototype.keypress = function(chars) {
      return this.state().keypress(this, chars.toString());
    };

    SyncPrompt.prototype.send_data = function() {
      var next, res;
      this.count++;
      next = (function(_this) {
        return function() {
          _this.lines = '';
          return _this.prompt();
        };
      })(this);
      res = typeof this.onData === "function" ? this.onData(this.lines.trim(), {
        next: next,
        stop: this.close
      }) : void 0;
      return (res != null ? typeof res.then === "function" ? res.then(next) : void 0 : void 0) || next();
    };

    SyncPrompt.prototype.prompt = function() {
      return this.state().prompt(this, ["[" + this.count + "] ", this.mode === 'js' ? chalk.white('pryjs') : chalk.blue('pryjs'), this.indent ? "* " + this.indent : '> '].join(''));
    };

    SyncPrompt.prototype.open = function() {
      var results;
      this.done = false;
      this.prompt();
      results = [];
      while (!this.done) {
        results.push(deasync.runLoopOnce());
      }
      return results;
    };

    SyncPrompt.prototype.type = function(input) {
      this.lines = input;
      return this.send_data();
    };

    SyncPrompt.prototype.close = function() {
      this.done = true;
      process.stdin.removeListener('data', this.keypress);
      return this.cli.close();
    };

    return SyncPrompt;

  })(EventEmitter);

  module.exports = SyncPrompt;

}).call(this);
